### DISCUSSION TITLE
More array shapes in the framework code

### DISCUSSION THREAD

**Discussion by tamiroh** [laravel/framework#48510] [General]
A few months ago I submitted [a PR that uses the array shape notation to write the return value in more detail](https://github.com/laravel/framework/pull/47759). 

```diff
      * Parse the route action into a standard array.
      *
      * @param  callable|array|null  $action
-     * @return array
+     * @return array{uses: callable|string, controller?: string, ...}
      *
      * @throws \UnexpectedValueException
      */
```

Unfortunately it was closed.

However, I think such detailed types are useful to improve the readability of the framework code. This is especially true where arrays are complicated, such as the above. In fact, [there are already several places in the framework where array-shape is used](https://github.com/search?q=repo%3Alaravel%2Fframework%20array%7B&type=code).

Any opinions on this?

I would like to bring the above PR to merge. Furthermore, if possible, I would like to apply the array shape typings to other types as well.

**Comment by henzeb** [comment]
I agree. But I suggest php 8 attributes for that.
